### PR TITLE
replace mlabwrap with python-matlab-bridge in Matlab tests

### PR DESCRIPTION
**Pull request by grlee77**
This PR replaces ``mlabwrap`` with [pymatbridge](https://github.com/arokem/python-matlab-bridge) to do the Matlab-based testing.  This is motivated primarily by the relative ease of installation for ``pymatbridge``and its compatibility with python 3.  ``pymatbridge`` is still being actively maintained.

This was discussed in #92 where @ThomasA brought up the extremely long runtime for the Matlab test suite.  Here the number of sizes tested has been greatly reduced (as discussed there) so that the total test time should be around 1-2 minutes.

The first commit implements the change to ``pymatbridge``.  The second reorders some of the loops to reduce the number of data transfers to Matlab and speed things up a bit more.

Please comment on whether the try/finally block as used here is the proper way to stop the Matlab session in the case of failure.

### CODE REVIEW

File: pywt/tests/test_matlab_compatibility.py [hunk 1]
```diff
@@ -13,11 +13,12 @@
...
-    from mlabwrap import mlab
+    from pymatbridge import Matlab
+    mlab = Matlab()
 except ImportError:
     print("To run Matlab compatibility tests you need to have MathWorks "
-          "MATLAB, MathWorks Wavelet Toolbox and mlabwrap Python extension "
-          "installed.")
+          "MATLAB, MathWorks Wavelet Toolbox and the pymatbridge Python "
+          "package installed.")
     _has_matlab = True
```
Comment (aaren) [inline pywt/tests/test_matlab_compatibility.py]: Can you rename `_has_matlab` to `_not_has_matlab` or similar? It's pretty confusing as is!

Comment (aaren) [inline pywt/tests/test_matlab_compatibility.py]: Also, move the initial declaration to the end of the `try:`

**Comment by aaren** [comment]
The try/finally is the correct way to ensure that matlab is shutdown when there is an error.

The other way of doing it would be to have a global `Matlab` instance started and stopped by setup and teardown functions. However, for the single test here I think try/finally is fine.

**Comment by aaren** [comment]
See the [numpy testing docs][npt] for info on setup / teardown.

[npt]: https://github.com/numpy/numpy/blob/master/doc/TESTS.rst.txt#easier-setup-and-teardown-functions--methods

**Comment by rgommers** [comment]
needs a rebase

**Comment by grlee77** [comment]
@aaren:  I believe all comments have been addressed and it has been rebased.
appveyor also [passed](https://ci.appveyor.com/project/rgommers/pywt/build/1.0.64)

I am not sure why commit 89ca2b6 appears first in the list of commits displayed above, when it should be the third commit.  Maybe I messed something up in the rebase?  (``git log`` on my local machine shows the commits in the expected order, not the one shown here)

**Comment by grlee77** [comment]
@rgommers 
I will add the ``dec.slow`` decorator as you suggest in #92.  Do you still want to use an environment variable such as PYWT_XSLOW that would modify to the full set of sizes that was there previously if it is set?
i.e. at the top of the file:
```
if 'PYWT_XSLOW' in os.environ:
    data_size = list(range(w.dec_len, 40)) + [100, 200, 500, 1000, 50000]
else:
    data_size = (w.dec_len, w.dec_len + 1)
```

**Comment by rgommers** [comment]
That depends on if the large number of values add anything. I'd think that if it's tested for say 5 sizes then it'll work for the other too. But I haven't looked into the detail and no access to Matlab to test this.

My rule of thumb is: a single test should be well under 1s to run in ``test()``, under 10-20s for ``test('full')`` and if it takes minutes we need an ``XSLOW`` envvar.

**Comment by aaren** [comment]
Let's keep the original list of sizes in there for now, using `XSLOW`. Might be there was an odd reason for them being there that has been forgotten.

@grlee77 do you have matlab access to test? I can probably sort something out if not.

**Comment by aaren** [comment]
@grlee77 r.e. commit order: are you doing `git push --force` when pushing from local to the issue branch?

**Comment by grlee77** [comment]
@aaren I did use ``--force`` after the rebase.

**Comment by grlee77** [comment]
@rgommers.  I have access to Matlab R2012.  It takes about 60-90 seconds to run the limited number of sizes.  To get it to 10-20 s we would have to also drop some of the modes or wavelet types, but I would be reluctant to do that.  I did not let the full set of sizes run to completion, but @ThomasA reported that it took about an hour.  I don't see any obvious reason to test that particular set of sizes.  Maybe one longer size of around 100 in addition to wav.dec_len and wav.dec_len+1 would be sufficient.

The other fast option would be to pre-store all Matlab outputs and compare those rather than running Matlab explicitly.

**Comment by ThomasA** [comment]
Maybe pre-storing Matlab outputs is actually a good idea. That should speed up the tests tremendously. 
BTW, I have access to the newest Matlab release and can run tests if you like.

**Comment by rgommers** [comment]
Storing might indeed make sense. If you can avoid creating object arrays (which won't be portable between py2 and py3 I just learned the hard way), then ``.npz`` format probably makes the most sense. Or plain text/csv files if those don't get too large. In ``.npz`` format, how large will the file be? Something on the order of a few Mb should be OK to check in. Much larger probably needs a bit of thought (at least stripping out of source releases).

**Comment by grlee77** [comment]
just verified that if running only the ``(w.dec_len, w.dec_len+1)`` sizes, the total number of floating point values to compare is 27128 (so < 2MB to store uncompressed doubles).  Python itself only takes about 30 ms to do the test, so that would be extremely fast.

The full set of sizes takes python only around 1.2s to run, but the data storage would be > 1GB, so clearly that is unreasonable to store.

**Comment by grlee77** [comment]
However, I think we should keep the pymatbridge option around as well in case we want to add other tests for 2d or 3d later.  Data sizes would quickly become unreasonable for many tests of those cases.

**Comment by rgommers** [comment]
So add ``(w.dec_len, w.dec_len+1)`` results as ``.npz`` and the rest behind ``XSLOW`` then?

**Comment by grlee77** [comment]
@rgommers, @aaren

Two new commits:
The first adds PYWT_XSLOW to enable the full set of sizes.  
The second adds the option to use a precomputed result. The keys into the .npz are based on the matlab mode name, wavelet name, and data size, so changes to any of those would require regenerating the precomputed results.

The data file size is 561,830 bytes, which is reasonable.

I thought it would be faster, but it takes 3-4 seconds on my machine even with the precomputed results, so I have left it marked as ``dec.slow``

**Comment by grlee77** [comment]
Also, if we decide to go with the precomputed option, should I store the script use to generate the results somewhere for reference?

**Comment by rgommers** [comment]
> Also, if we decide to go with the precomputed option, should I store the script use to generate the results somewhere for reference?

That would be useful I think, just adding it in ``tests/data/`` would be a good option.

**Comment by aaren** [comment]
@grlee77 good work. Looks good to me now I think.

`@dec.slow` tests get run by travis, right?

**Comment by rgommers** [comment]
> ``@dec.slow`` tests get run by travis, right?

Yes, they get run for Python 2.7. They're run with ``pywt.test('full')``

**Comment by rgommers** [comment]
Passed https://ci.appveyor.com/project/rgommers/pywt/build/1.0.67 as well. I really should fix that asap.

**Comment by rgommers** [comment]
Merged, thanks @grlee77 and @aaren

**Pull request merged**